### PR TITLE
Onboard to central package management in NuGet.Client repo

### PR DESCRIPTION
**Pull request by jeffkl**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1969

Regression? Last working version:

## Description

* I tried using `git mv` on `packages.targets` but I changed too much of the contents so it looks like a delete/add 😢 
* Organized all of the package versions in `Directory.Packages.props`
* Added `GlobalPackageReference` items for `PackageReference` items that were in common imports
* Used `<PackageReference />` instead of `<PackageDownload />` since CPM doesn't support PackageDownload.  This brought in new dependencies which weren't being pulled in before so I had to resolve all of the conflicts with a few new package versions and/or `PrivateAssets`/`ExcludeAssets`.
* Removed logic that verified package versions were declared in one place since CPM does that
* Used `GeneratePathProperty` in a few places
* Had to include some interfaces for `project.json` in `Test.Utility` since I think they were removed from newer versions of `Microsoft.VisualStudio.ProjectSystem.Managed`.  They are just using `ComImport` so it should be okay.
* Writing an empty `Directory.Packages.props` to the `.test` folder

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: Directory.Packages.props
Comment (jeffkl) [inline Directory.Packages.props:143]: Note to self: Check if `UsePublicApiAnalyzer` is used anywhere.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectHelper.cs
Comment (jeffkl) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectHelper.cs:29]: Should I put all required source code changes in their own commit for easier review?

File: test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/NuGet.XPlat.FuncTest.csproj
Comment (jeffkl) [inline test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/NuGet.XPlat.FuncTest.csproj:7]: Note to self: Check if this NoWarn is still needed

File: test/TestUtilities/Test.Utility/ProjectManagement/TestProjectKProject.cs
Comment (jeffkl) [inline test/TestUtilities/Test.Utility/ProjectManagement/TestProjectKProject.cs:78]: Note to self: Maybe place these in a different file and explain why they're needed?

File: Directory.Packages.props
Comment (zivkan) [inline Directory.Packages.props:114]: childish rant: I still don't get what benefit this has over `<PackageReference`. I guess you don't need to also specify a `PackageVersion`, so it's one less line? 🤷

File: src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj
Comment (zivkan) [inline src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj]:
```suggestion
    <PackageReference Include="NuGet.Core" ExcludeAssets="All" GeneratePathProperty="true" />
```

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj]: Looks like this assembly reference could be removed if we change the `PackageReference` to have `ExludeAssets="runtime"` instead of including `None`.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:42]: I don't think this is an improvement over `PackageDownload` with assembly references. In fact, it's one line longer, despite the `<Reference` items being 3 lines each.

File: test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/NuGet.XPlat.FuncTest.csproj
Comment (zivkan) [inline test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/NuGet.XPlat.FuncTest.csproj:32]: Every single time I tried to improve NuGet.XPlat.FuncTest, I ran into endless problems with all this compiler package and Microsoft.CSharp.Core.targets file stuff.  Good work finding something that worked!👏

I still hope one day we can figure out why it's needed, and find a way to make tests not need it, but one step at a time.

File: test/TestExtensions/GenerateLicenseList/GenerateLicenseList.csproj
Comment (zivkan) [inline test/TestExtensions/GenerateLicenseList/GenerateLicenseList.csproj]: Did you try removing the version override, and see if there are still problems? I don't think there are any good reasons for this project to use older analyers.

**Review by zivkan — DISMISSED** [review]
I don't like the "enumerate every dependency of package X and add a `<PackageReference Inlcude="x" ExcludeAssets="all" />` as an alternative to `<PackageDownload`.  But everything else is opinionated, so if others approve I won't complain if it gets merged as is.

File: Directory.Packages.props
Comment (jeffkl) [reply Directory.Packages.props:114]: Its really just syntactic sugar which helps you get a package reference correct when its in a common import.  Think of it like `using()` which the compiler just turns into `try { } finally { obj.Dispose(); }`.  Either pattern works but one is the pit of success.

File: src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj
Comment (jeffkl) [reply src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj]: Do you think this needs a `PrivateAssets="All"` still?

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:42]: I'm not a huge fan either. 
I think we should either reference a package version and let everything be done automatically by NuGet or keep using PackageDownload. 

Why do we have so many conflicts? Maybe we should talk/help the project system to fix those conflicts if they are on their end?

I know it's keeping things in multiple places, but I think the natural request is that PackageDownload is supported with CPM eventually in some way shape of form.

File: test/TestExtensions/GenerateLicenseList/GenerateLicenseList.csproj
Comment (nkolev92) [reply test/TestExtensions/GenerateLicenseList/GenerateLicenseList.csproj]: There are breaking API changes. This project generates code, so uses APIs from these packages, idk how complicated fixing that would be. 

We should at least track it if we want to make it consistent eventually, but imo given that this is something that just generates some minimal amounts of code, I don't care too much about it using an updated version.

File: Directory.Packages.props
Comment (nkolev92) [inline Directory.Packages.props:118]: We probably need to wait to merge this until we know 100% all the CIs have 17.4. 

We should also update our requirements section for building the repo.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj
Comment (jeffkl) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj]: Yup, good catch.  It seems to work so I'll update this PR

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj
Comment (jeffkl) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:42]: These are the conflicts we get if we don't elevate transitive references to top-level:

```
error NU1603: Warning As Error: Microsoft.VisualStudio.ProjectSystem 17.4.221-pre depends on Microsoft.Build (>= 17.3.0-preview-22328-04) but Microsoft.Build 17.3.0-preview-22328-04 was not found. An approximate best match of Microsoft.Build 17.3.0 was resolved.
error NU1603: Warning As Error: Microsoft.VisualStudio.ProjectSystem 17.4.221-pre depends on Microsoft.Build.Tasks.Core (>= 17.3.0-preview-22328-04) but Microsoft.Build.Tasks.Core 17.3.0-preview-22328-04 was not found. An approximate best match of Microsoft.Build.Tasks.Core 17.3.0 was resolved.
error NU1603: Warning As Error: Microsoft.VisualStudio.ProjectSystem 17.4.221-pre depends on Microsoft.Build.Utilities.Core (>= 17.3.0-preview-22328-04) but Microsoft.Build.Utilities.Core 17.3.0-preview-22328-04 was not found. An approximate best match of Microsoft.Build.Utilities.Core 17.3.0 was resolved.
error NU1603: Warning As Error: Microsoft.VisualStudio.ProjectSystem.Managed 17.2.0-beta1-20502-01 depends on Microsoft.CodeAnalysis (>= 4.1.0-2.21558.8) but Microsoft.CodeAnalysis 4.1.0-2.21558.8 was not found. An approximate best match of Microsoft.CodeAnalysis 4.1.0-3.final was resolved.
error NU1603: Warning As Error: Microsoft.VisualStudio.ProjectSystem.Managed.VS 17.2.0-beta1-20502-01 depends on Microsoft.CodeAnalysis (>= 4.1.0-2.21558.8) but Microsoft.CodeAnalysis 4.1.0-2.21558.8 was not found. An approximate best match of Microsoft.CodeAnalysis 4.1.0-3.final was resolved.
error NU1102: Unable to find package Microsoft.CodeAnalysis.Features with version (= 4.1.0-2.21558.8)
error NU1102:   - Found 3002 version(s) in myget-legacy@Local [ Nearest version: 3.6.0-beta1-20111-10 ]
error NU1102:   - Found 130 version(s) in dotnet-public [ Nearest version: 4.1.0-3.final ]
error NU1102:   - Found 1 version(s) in vside_vs-impl [ Nearest version: 2.6.0-vs-for-mac-62317-08 ]
error NU1102:   - Found 0 version(s) in vside_vssdk
error NU1102:   - Found 0 version(s) in dotnet-eng
error NU1102:   - Versions from nuget-build were not considered
error NU1107: Version conflict detected for Microsoft.CodeAnalysis.Common. Install/reference Microsoft.CodeAnalysis.Common 4.1.0-3.final directly to project NuGet.VisualStudio.Implementation to resolve this issue.
error NU1107:  NuGet.VisualStudio.Implementation -> Microsoft.VisualStudio.ProjectSystem.Managed.VS 17.2.0-beta1-20502-01 -> Microsoft.CodeAnalysis 4.1.0-3.final -> Microsoft.CodeAnalysis.VisualBasic.Workspaces 4.1.0-3.final -> Microsoft.CodeAnalysis.Common (= 4.1.0-3.final)
error NU1107:  NuGet.VisualStudio.Implementation -> Microsoft.VisualStudio.ProjectSystem.Managed 17.2.0-beta1-20502-01 -> Microsoft.VisualStudio.LanguageServices 4.1.0-2.21558.8 -> Microsoft.CodeAnalysis.Common (= 4.1.0-2.21558.8).
  Failed to restore D:\NuGet\src\NuGet.Clients\NuGet.VisualStudio.Implementation\NuGet.VisualStudio.Implementation.csproj (in 1.01 sec).
```

1. `PackageDownload` doesn't support `GeneratePathProperty` so you need to declare a property for the `PackageReference` and the `Reference` to keep things in sync.  I personally don't find that to be better.
2. `PackageDownload` doesn't resolve transitive dependencies so I'm guessing it was only used here as a way of getting around the package conflicts.  That seems like a hack?
3. One of the references aren't even added as a `PackageReference`.  Which means we are just relying on the fact that some other project is putting that package in our global packages folder.  Using PackageReferences in a stock way helps avoid that.
  ```xml
<Reference Include="Microsoft.VisualStudio.ProjectSystem.VS">
    <HintPath>$(NuGetPackageRoot)\microsoft.visualstudio.projectsystem\16.7.156-pre\lib\net472\Microsoft.VisualStudio.ProjectSystem.VS.dll</HintPath>
</Reference>
  ```
4. I certainly am not focused on number of lines of XML here.  My focus was to use NuGet in a stock way and resolve package conflicts so that our product builds.  The fact that I had to add all of this speaks to the customer pain that people go through all of the time.  Isn't it a good thing for us to experience it in our own repo rather than work around it?

File: test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/NuGet.XPlat.FuncTest.csproj
Comment (jeffkl) [reply test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/NuGet.XPlat.FuncTest.csproj:32]: I probably got lucky that packages exist now with the assets we need for .NET Framework and .NET Core.  From what I can tell, we just really need `Microsoft.CSharp.Core.targets`.  We could probably just copy that out of the SDK at runtime but do that in a future change.

File: test/TestExtensions/GenerateLicenseList/GenerateLicenseList.csproj
Comment (jeffkl) [reply test/TestExtensions/GenerateLicenseList/GenerateLicenseList.csproj]: It used to give package conflicts because we were referencing an older version of `Microsoft.CodeAnalysis` but now I've updated us to `4.3.1` and it seems to fix the problem.  So Andy is right that I can remove this now...

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj
Comment (jeffkl) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:42]: What's the consensus here?  Should I have `<PackageDownload Update="" />` in `Directory.Build.targets` or should I use CPM like this?

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:42]: I think a way I can summarize this is that GeneratePathProperty and PackageDownload are both hacks. Ideally we just a normal PackageReference. Going from hack X to hack Y is not enough. I'm hoping we can remove the hacks.

> PackageDownload doesn't support GeneratePathProperty so you need to declare a property for the PackageReference and the Reference to keep things in sync. I personally don't find that to be better.

GeneratePathProperty for PackageDownload is on the backlog https://github.com/NuGet/Home/issues/8476. 
IMO, as long as you're doing something manual, you're failing. I don't find using 1 property instead of 2 is any better. 

We should aim to *not* use GeneratePathProperty. 

> PackageDownload doesn't resolve transitive dependencies so I'm guessing it was only used here as a way of getting around the package conflicts. That seems like a hack?

Yes, in general any usage of ExcludeAssets=all or GeneratePathProperty is also a hack, a sign that a package was either authored poorly or we have a weird shipping method.

Yes, I'd argue the same thing applies to specifying ExcludeAssets everywhere and using GeneratePathProperty.

> One of the references aren't even added as a PackageReference. Which means we are just relying on the fact that some other project is putting that package in our global packages folder. Using PackageReferences in a stock way helps avoid that.
 
 It relies on PackageDownload to get the package no?

> I certainly am not focused on number of lines of XML here. My focus was to use NuGet in a stock way and resolve package conflicts so that our product builds. The fact that I had to add all of this speaks to the customer pain that people go through all of the time. Isn't it a good thing for us to experience it in our own repo rather than work around it?

I think the customer pain here is because of an incorrect authored package. Similarly to what our customers would do, I think we should work with the package publishers to help us remove the hacks.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:42]: I don't want to block this PR on that work ofc. Mostly just saying that this is a lateral move, rather than an improvement.

File: src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj
Comment (zivkan) [reply src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj]: I'm not 100% sure what the implementation is, but to me `IncludeAssets` means "use in current project and flow transitively", `PrivateAssets` means "use in current project, but don't flow transitively", and `ExcludeAssets means "don't use in current project, or flow transitively".

So no, I don't believe that `PrivateAssets` are needed.

File: test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/NuGet.XPlat.FuncTest.csproj
Comment (zivkan) [reply test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/NuGet.XPlat.FuncTest.csproj:32]: I'm really confused why `Microsoft.CSharp.core.targets` is needed at all.  What are the tests doing that actually import that file? 🤷‍♂️    anyway, way off topic for this PR. I'm just happy that it didnt' block you like it's blocked me so many times before.

**Review by zivkan — DISMISSED** [review]

File: src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj]: NuGet.Commandline is a special package that suppresses all dependencies, https://github.com/NuGet/NuGet.Client/blob/a487c8336b29a0e53dc46f4aab424fdd5de50e72/src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj#L20.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:42]: I took a look and this is the only change leftover that I'm interested in. 

If you 2 prefer this approach(@zivkan approved), feel free to keep it (ideally add a comment letting me know :D)  Happy to have an offline chat about it if needed.

Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:42]: I'd prefer to keep the old `<PackageDownload`. We did it specifically because we want compile time APIs only. We don't run tests against their APIs, so we don't need their dependencies. We don't even need their dll in our test's bin folder.

But from a pragmatic point of view, it doesn't matter so much. I think the difference is with `PackageDownload` when we have to increment the package version, it's one or two places we need to do it. With PackageReference, there's a whole lot extra work if project system's dependencies change. But that can always be the problem of the person who needs to bump the package version.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:42]: We can leave it be then.

We should really talk to the project-system team and get them to publish their dependencies as well.

**Comment by jeffkl** [comment]
> I forgot about the comment about changing the repo readme.md or contribution.md to mention minimum VS version numbers required to build the repo

Good idea, I have updated the `CONTRIBUTING.md` to indicate you need VS 17.4+ and .NET SDK 7.0+

**Review by nkolev92 — APPROVED** [review]

**Review by zivkan — APPROVED** [review]

**Pull request merged**
